**************************
List of integrated sources
**************************

RNA-KG is a knowledge graph encompassing biological knowledge about RNAs gathered from more than 50 public databases. Here, we list integrated sources and ontologies. Special thank to `DrugBank <https://go.drugbank.com/>`_ for liking this project and making me access supplementary data!

- RNA sources:
    - miRNA Databases:
        - `miRBase <https://www.mirbase.org/>`_
        - `miRDB <https://mirdb.org/>`_
        - `miRNet <https://www.mirnet.ca/miRNet>`_
        - `miRecords <http://c1.accurascience.com/miRecords/download_data.php?v=4>`_
        - `EpimiR <http://www.jianglab.cn/EpimiR/index.jsp>`_
        - `HMDD <https://www.cuilab.cn/hmdd>`_
        - `miR2Disease <http://watson.compbio.iupui.edu:8080/miR2Disease/>`_
        - `TargetScan <https://www.targetscan.org/vert_80/>`_
        - `SomamiR DB <https://compbio.uthsc.edu/SomamiR/>`_
        - `TarBase <https://dianalab.e-ce.uth.gr/html/diana/web/index.php?r=tarbasev8/index>`_
        - `miRTarBase <https://mirtarbase.cuhk.edu.cn/~miRTarBase/miRTarBase_2022/php/index.php>`_
        - `SM2miR <http://www.jianglab.cn/SM2miR/>`_
        - `TransmiR <https://www.cuilab.cn/transmir>`_
        - `PolymiRTS <https://compbio.uthsc.edu/miRSNP/>`_
        - `dbDEMC <https://www.biosino.org/dbDEMC/index>`_
        - `TAM <http://www.lirmed.com/tam2/>`_
        - `PuTmiR <https://www.isical.ac.in/~bioinfo_miu/TF-miRNA1.php>`_
        - `miRPathDB <https://mpd.bioinf.uni-sb.de/overview.html>`_
        - `miRCancer <http://mircancer.ecu.edu/>`_
        - `miRdSNP <http://mirdsnp.ccr.buffalo.edu/index.php>`_
        - `miRandola <http://mirandola.iit.cnr.it/index.php>`_

    - mRNA Vaccine Database:
        - `DrugBank <https://go.drugbank.com/categories/DBCAT005631>`_

    - siRNA Databases:
        - `ICBP siRNA <http://web.mit.edu/sirna/index.html>`_
        - `DrugBank <https://go.drugbank.com/categories/DBCAT005484>`_

    - RNA Aptamer Databases:
        - `Apta-Index <https://www.aptagen.com/apta-index/>`_
        - `DrugBank <https://go.drugbank.com/categories/DBCAT001641>`_

    - ASO Databases:
        - `eSkip-Finder <https://eskip-finder.org/cgi-bin/input.cgi>`_
        - `DrugBank <https://go.drugbank.com/categories/DBCAT001709>`_

    - gRNA Database:
        - `Addgene <https://www.addgene.org/>`_

    - lncRNA Databases:
        - `LncBook <https://ngdc.cncb.ac.cn/lncbook/>`_
        - `LncRNADisease <http://www.rnanut.net/lncrnadisease/>`_
        - `LncExpDB <https://ngdc.cncb.ac.cn/lncexpdb/>`_
        - `dbEssLnc <https://esslnc.pufengdu.org/home>`_
        - `lncATLAS <https://lncatlas.crg.eu/>`_
        - `NONCODE <http://www.noncode.org/index.php>`_
        - `Lnc2Cancer <http://bio-bigdata.hrbmu.edu.cn/lnc2cancer/>`_
        - `LncRNAWiki <https://ngdc.cncb.ac.cn/lncrnawiki/>`_

    - Ribozyme Database:
        - `Ribocentre <https://www.ribocentre.org/>`_
        - `Rfam <http://rfamlive.xfam.org/>`_

    - Viral RNA Database:
        - `ViroidDB <https://viroids.org/>`_

    - Riboswitch Databases:
    	- `TBDB <https://tbdb.io/>`_
    	- `RSwitch <https://penchovsky.atwebpages.com/applications.php?page=58>`_

    - tRF Databases:
    	- `tRFdb <http://genome.bioch.virginia.edu/trfdb/index.php>`_
    	- `tsRFun <https://rna.sysu.edu.cn/tsRFun/index.php>`_
    	- `MINTbase <https://cm.jefferson.edu/MINTbase/>`_

    - snoRNA Database:
    	- `snoDB <https://bioinfo-scottgroup.med.usherbrooke.ca/snoDB/>`_

    - tRNA Database:
    	- `tRNAdb <http://trna.bioinf.uni-leipzig.de/DataOutput/>`_

    - InterRNA Databases:
    	- `RNAInter <http://www.rnainter.org/>`_
    	- `RNALocate <http://www.rna-society.org/rnalocate/>`_
    	- `RNADisease <http://www.rnadisease.org/>`_
    	- `ncRDeathDB <https://www.rna-society.org/ncrdeathdb/>`_
    	- `cncRNADB <http://www.rna-society.org/cncrnadb/>`_
    	- `ViRBase <http://www.rna-society.org/virbase/>`_
    	- `Vesiclepedia <http://microvesicles.org/>`_
    	- `DirectRMDB <www.rnamd.org/directRMDB/index.html>`_
    	- `Modomics <https://genesilico.pl/modomics/>`_

- Ontologies:
    	- `Gene Ontology (GO) <https://geneontology.org/>`_
        - `Mondo Disease Ontology (Mondo) <https://mondo.monarchinitiative.org/>`_
        - `Vaccine Ontology (VO) <https://violinet.org/vaccineontology/>`_
        - `Chemical Entities of Biological Interest (ChEBI) <https://www.ebi.ac.uk/chebi/>`_
        - `Uber-anatomy ontology (Uberon) <http://obophenotype.github.io/uberon/>`_
        - `Cell Line Ontology (CLO) <http://www.clo-ontology.org/>`_
        - `PRotein Ontology (PRO) <https://proconsortium.org/>`_
        - `Sequence Ontology (SO) <http://www.sequenceontology.org/>`_
        - `Pathway Ontology (PW) <https://rgd.mcw.edu/rgdweb/ontology/search.html>`_

------------

For standardizing metadata, we considered:
    - `dbSNP <https://www.ncbi.nlm.nih.gov/snp/>`_
    - `PubMed <https://pubmed.ncbi.nlm.nih.gov/>`_
    - `NCI Thesaurus OBO Edition (NCIT) <https://github.com/NCI-Thesaurus/thesaurus-obo-edition>`_
    - `Chemical Methods Ontology (CHMOD) <https://github.com/rsc-ontologies/rsc-cmo>`_
